File: shared.js
import { parse } from "path";

export const headers = {
    PING: "udpensor:ping",
    PONG: "udpensor:pong",
    FEED: "udpensor:feed"
}

/**
 * 
 * @param {number} seconds 
 */
export const getSeconds = seconds => seconds * 1000;

/**
 * 
 * @param {number} minutes 
 */
export const getMinutes = minutes => getSeconds(minutes * 60);

/**
 * 
 * @param  {string[]} args 
 */
export const groupArguments = (args) => {
    return args.reduce((prev, next, i) => {
        if (i < 2) return prev;
        if (next.startsWith('-')) {
            prev.push([next]);
            return prev;
        }
        prev[prev.length - 1].push(next);
        return prev;
    }, []);
}

/**
 * 
 * @param {string[][]} groups 
 * @param {number} defaultPort
 */
export const parseArguments = (args, defaultPort) => {
    const groups = groupArguments(args);
    const parsed = {
        ports: [],
        bindPort: -1,
        broadcastIP: ""
    }
    for (const group of groups) {
        if (["-p", "--port", "--ports"].includes(group[0])) {
            parsed.ports.push(...group.slice(1));
        }
        if (["-bp", "--bind-port"].includes(group[0])) {
            parsed.bindPort = group.length > 1 ? group[1] : defaultPort;
        }
        if (["-bip", "--broadcast-ip"].includes(group[0])) {
            parsed.broadcastIP = group.length > 1 ? group[1] : "255.255.255.255";
        }
        // TODO: add other arguments
    }
    if (parsed.bindPort === -1) { parsed.bindPort = defaultPort; }
    return parsed;
}

File: Shared/Shared.fs
module Shared

open System

type Headers =
    static member PING = "udpensor:ping"
    static member PONG = "udpensor:pong"
    static member FEED = "udpensor:feed"


let getMlsFromSeconds (seconds: int) =
    TimeSpan.FromSeconds(seconds |> float).TotalMilliseconds
    |> int

let getMlsFromMinutes (minutes: int) =
    TimeSpan.FromMinutes(minutes |> float).TotalMilliseconds
    |> int

let private groupArguments (args: seq<string>) =
    args
    |> Seq.skip (2)
    |> Seq.fold (fun (prev: ResizeArray<ResizeArray<string>>) (next: string) ->
        if next.StartsWith('-') then
            let item = ResizeArray()
            item.Add(next)
            prev.Add(item)
        else 
            prev.[prev.Count - 1].Add(next)
        prev) (ResizeArray<ResizeArray<string>>())
    |> List.ofSeq

let parseArguments (args: seq<string>) (defaultPort: int) (defautBroadcastIP: string) =
    let groups = groupArguments args

    let mutable parsed =
        {| ports = List.empty<string>
           bindPort = None
           broadcastIP = None |}

    for group in groups do
        let group = group |> Seq.toList
        if [ "-p"; "--port"; "--ports" ]
           |> List.contains (group.Head) then
            parsed <- {| parsed with ports = group.Tail|}
        if [ "-bp"; "--bindPort"; "--bind-port" ]
           |> List.contains (group.Head) then
            parsed <-
                {| parsed with
                       bindPort = Some(if group.Length > 1 then group.[1] |> int else defaultPort) |}
        if [ "-bip"
             "--broadcastIP"
             "--broadcast-ip" ]
           |> List.contains (group.Head) then
            parsed <-
                {| parsed with
                       broadcastIP = Some(if group.Length > 1 then group.[1] else defautBroadcastIP) |}

    // TODO: add other arguments

    {| parsed with
           bindPort = parsed.bindPort |> Option.defaultValue defaultPort
           broadcastIP =
               parsed.broadcastIP
               |> Option.defaultValue defautBroadcastIP |}
